### PR TITLE
Feature/repository

### PR DESCRIPTION
**Pull request by gkellogg**
This should be a fairly functionally complete Repository class with support for search and query. Note that it does need to use SPARQL 1.1 features, principally for the aggregation necessary for doing `facet_counts`. There are open issues with FIXME where it's not clear how much of the Solr model needs to be replicated.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -59,7 +59,9 @@ gem 'sparql',         '~> 1.99'
...
 group :development, :test do
-  gem 'rdf-do',       '~> 1.99' # Using DataObjects for SPARQL store
-  gem 'data_objects', '~> 0.10'
-  gem 'do_sqlite3' ,  '~> 0.10'
+  #gem 'rdf-do',       '~> 1.99' # Using DataObjects for SPARQL store
+  #gem 'data_objects', '~> 0.10'
+  #gem 'do_sqlite3' ,  '~> 0.10'
+  gem 'rdf-mongo'
+  gem 'bson_ext'
 end
```
Comment (darrenleeweber) [inline Gemfile]: There might be problems running specs on travis CI with a mongo dependency.  In general, mongo is a reasonable choice and one that I am open to supporting, although it's not widely adopted in DLSS and the dev-ops team have not committed to supporting it in production (someone please correct me if that's wrong).

**Comment by darrenleeweber** [comment]
Difficult to comment on changes to anything in `app/models/blacklight/sparql/` and associated specs without reviewing the entire file contents.  That's not necessary, I trust @gkellogg with this stuff ;-)

File: Gemfile (hunk 1 above)
Comment (cbeer) [inline Gemfile]: @darrenleeweber I don't think we intend to run this code in production; the intent is only to produce a proof-of-concept implementation for the discovery of linked data resources, and then consider extracting the good bits to a blacklight plugin. The mongo dependency here is purely incidental .

Comment (darrenleeweber) [inline Gemfile]: Thanks for clarification and that makes sense.  I'd be interested in some general enlightenment about the architecture considerations - curious about this approach and generally how an RDF/SPARQL approach can provide additional services that Solr can't provide.  I'm assuming this tackles problems that are not related to just pushing triple store data into Solr (or a third party sync, like FC4 via messaging).

Comment (gkellogg) [inline Gemfile]: Mongo can run on Travis without approved, the rdf-mongo gem is tested on Travis.

I tried running using rdf-do on sqlite3, but it is unbearably slow. Tests could probably run using the memory-hash RDF::Repository implementation, but not useful for general development.

Comment (gkellogg) [inline Gemfile]: @cbeer is correct, in production, it would point to an external SPARQL store.

Comment (cbeer) [inline Gemfile]: > curious about this approach and generally how an RDF/SPARQL approach can provide additional services that Solr can't provide

I think that's exactly what we're setting out to discover.

**Comment by gkellogg** [comment]
It would be useful to check the expected inputs on outputs of the Repository methods. I'm uncertain about how search_builder creates the parameters for Repository#search, and the general format of documents and facet_counts in the Response.

**Comment by darrenleeweber** [comment]
@gkellogg - I don't have time to review all the details, but from what I've seen, I'm OK with merging this.  I'm not familiar with details of blacklight, so I'll defer to Chris, Jack or Jessie on that.  Please ping me if there's any specific RDF/SPARQL stuff to review.

**Comment by mjgiarlo** [comment]
:clap:

**Pull request merged**